clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2167,15 +2358,25 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     return true;
   }
 
-  bool handleNonDecompStruct(const CXXRecordDecl *, FieldDecl *FD,
+  bool handleNonDecompStruct(const CXXRecordDecl *RD, FieldDecl *FD,
                              QualType Ty) final {
-    addParam(FD, Ty);
+    CXXRecordDecl *FieldRecordDecl = Ty->getAsCXXRecordDecl();
+    SyclKernelPointerHandler PointerHandler(SemaRef, FieldRecordDecl);
+    KernelObjVisitor Visitor{SemaRef};
```

Comment:
Probably a comment explaining why we are visiting here again will be helpful.

Reply:
Yes will do. Sorry I thought I did. I'll explain the logic a bit more in comments and modify the tests so you can see what this PR is actually doing.